### PR TITLE
Fix to use own validator for Header Accept

### LINKED ISSUE #431: Unable to use own validator for header Accept
Since the update from 2.1.0 to 2.2.0 it is not possible anymore to have own validation for the header Accept (Content negotiation).

Example code I use, where I use `_accept` or replace it with a tuple `('text/json', 'application/json')` or a list `['text/json', 'application/json']`:
```
def _accept(request):
    return ('text/json', 'application/json')

@resource(collection_path='/items', path='/items/{item}', accept=_accept)
class Items(object):
    def __init__(self, request):
        self.request = request
    def collection_get(self):
        return {'abc': 'def'}
   ....
```
The error always occurs at same code section, only the underlying error changes depending on the argument:
1. With a callable: `AttributeError: 'function' object has no attribute 'lower'`
2. With a tuple: `AttributeError: 'tuple' object has no attribute 'lower'
3. With a list: `AttributeError: 'list' object has no attribute 'lower'

Backtrace:
```2017-01-04 00:02:36,025 ERROR [waitress][waitress] Exception when serving /items
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/waitress/channel.py", line 338, in service
    task.service()
  File "/usr/local/lib/python2.7/dist-packages/waitress/task.py", line 169, in service
    self.execute()
  File "/usr/local/lib/python2.7/dist-packages/waitress/task.py", line 399, in execute
    app_iter = self.channel.server.application(env, start_response)
  File "/usr/local/lib/python2.7/dist-packages/pyramid/router.py", line 236, in __call__
    response = self.invoke_subrequest(request, use_tweens=True)
  File "/usr/local/lib/python2.7/dist-packages/pyramid/router.py", line 211, in invoke_subrequest
    response = handle_request(request)
  File "/usr/local/lib/python2.7/dist-packages/pyramid/tweens.py", line 51, in excview_tween
    request_iface=request_iface.combined
  File "/usr/local/lib/python2.7/dist-packages/pyramid/view.py", line 547, in _call_view
    response = view_callable(context, request)
  File "/usr/local/lib/python2.7/dist-packages/pyramid/viewderivers.py", line 413, in viewresult_to_response
    result = view(context, request)
  File "/usr/local/lib/python2.7/dist-packages/pyramid/tweens.py", line 22, in excview_tween
    response = handler(request)
  File "/usr/local/lib/python2.7/dist-packages/pyramid/router.py", line 83, in handle_request
    info = routes_mapper(request)
  File "/usr/local/lib/python2.7/dist-packages/pyramid/urldispatch.py", line 95, in __call__
    if preds and not all((p(info, request) for p in preds)):
  File "/usr/local/lib/python2.7/dist-packages/pyramid/urldispatch.py", line 95, in <genexpr>
    if preds and not all((p(info, request) for p in preds)):
  File "/usr/local/lib/python2.7/dist-packages/pyramid/config/predicates.py", line 141, in __call__
    return self.val in request.accept
  File "/usr/local/lib/python2.7/dist-packages/webob/acceptparse.py", line 112, in __contains__
    if self._match(mask, offer):
  File "/usr/local/lib/python2.7/dist-packages/webob/acceptparse.py", line 290, in _match
    if (mask.lower() == offer.lower() or
AttributeError: 'function' object has no attribute 'lower'
```
If I use a string to specify validation (e.g. "text/json") and then call the URL with a different Accept header, cornice doesn't response with http status of `406 Not Acceptable`. Now it respond with `404 Not Found`:
```
curl -vvv http://localhost:43043/items -H "Accept: application/json"
*   Trying ::1...
* TCP_NODELAY set
* connect to ::1 port 43043 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 43043 (#0)
> GET /items HTTP/1.1
> Host: localhost:43043
> User-Agent: curl/7.52.1
> Accept: application/json
>
< HTTP/1.1 404 Not Found
< Content-Length: 58
< Content-Type: text/plain; charset=UTF-8
< Date: Wed, 04 Jan 2017 00:06:58 GMT
< Server: waitress
< X-Content-Type-Options: nosniff
<
404 Not Found

The resource could not be found.


/items
```

### PR DESCRIPTION
**Pull request by j-carl**
fixes #431

### CODE REVIEW

File: tests/test_resource.py [hunk 1]
```diff
@@ -19,12 +19,37 @@
...
     return [(Allow, 'alice', 'read')]
 
 
+def _accept(request):
+    return ('text/json', 'application/json')
+
+
+@resource(collection_path='/fruits', path='/fruits/{id}',
+          name='fruit_service', accept=_accept)
+class Group(object):
```
Comment (leplatrem) [inline tests/test_resource.py]: nitpick: `Group` versus `/fruits`

File: tests/test_resource.py [hunk 2]
```diff
@@ -19,12 +19,37 @@
...
+
+    def __init__(self, request, context=None):
+        self.request = request
+        self.context = context
+
+    def collection_get(self):
+        return {'fruits': list(FRUITS.keys())}
```
Comment (leplatrem) [inline tests/test_resource.py]: it does not seem to be used in tests?

File: tests/test_resource.py [hunk 3]
```diff
@@ -19,12 +19,37 @@
...
+
+    @view(renderer='jsonp')
+    @view(renderer='json')
+    def get(self):
+        return FRUITS.get(int(self.request.matchdict['id']))
```
Comment (leplatrem) [inline tests/test_resource.py]: ditto: it does not seem to be used in tests?

File: tests/test_resource.py [hunk 4]
```diff
@@ -125,6 +150,17 @@ def test_accept_headers(self):
...
+    def test_accept_headers_callable(self):
+        self.assertEqual(
+            self.app.post("/fruits", headers={'Accept': 'text/json'},
+                          params=json.dumps({'test': 'yeah'})).json,
+            {'test': 'yeah'})
+
+        self.assertEqual(
+            self.app.post("/fruits", headers={'Accept': 'application/json'},
+                          params=json.dumps({'test': 'yeah'})).json,
+            {'test': 'yeah'})
```
Comment (leplatrem) [inline tests/test_resource.py]: We may want to add a call that fails with a 415 to make sure we actually use what `_accepts()` returns

File: cornice/pyramidhook.py [hunk 5]
```diff
@@ -204,6 +204,11 @@ def register_service_views(config, service):
 
     routes = config.get_predlist('route')
     for predicate in routes.sorter.names:
+        # Do not let the custom predicates handle the Header Accept
+        # it is done in the fallback view
```
Comment (leplatrem) [inline cornice/pyramidhook.py]: I have trouble understanding why this change allows to use a callable for `accept`. I know I'm not extremely familiar with resources but this comment could be improved maybe? At least with a link to the related issue to help the reader.

**Comment by j-carl** [comment]
Just to clarify things a little: The error also happens when you use services. It just happened to me when I was working with resources and that's the reason I wrote the tests against resources.

I looked in the tests for services and validation but couldn't figure out where to place things (I know it's not the ideal solution, but the resource tests are also covering the underlying problem).

**Comment by leplatrem** [comment]
Sorry for the delay!

**Pull request merged**